pkg/spiffecert/writer.go

```diff
@@ -0,0 +1,118 @@
+// Package spiffecert contains cert reload and spiffe items
+// Copyright (c) 2016-2021, F5 Networks, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package spiffecert
+
+import (
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+)
+
+const (
+	bundleFileMode = os.FileMode(0o644)
+	certsFileMode  = os.FileMode(0o644)
+	keyFileMode    = os.FileMode(0o600)
+)
+
+//go:generate counterfeiter -generate
+
+// DiskSVIDConfig contains the configuration for a Writer.
+type DiskSVIDConfig struct {
```

Comment:
Was this the old code that we used to use in the sidecar for writing to disk? But I'm guessing still needed for NIC?

Reply:
It's also a little odd because we have the CABundleManager which writes the CA to disk, and then the DiskSvidWriter which writes everything to disk.

Is there a way for us to collapse the common pieces of this?

Reply:
Disksvidwriter is used in the following places:
- cert-reloader to save certs to disk
- unit tests for certmanager
- mesh-api to save current certs to disk

Regarding splitting the SVID writers into separate CA writers and Cert/Key writers I can only see savings of a couple of lines around just writing the CA. In fact, the CABundleManager has its own special logic handling minimizing writes. Meanwhile splitting it into separate writers for the CA and Key/Cert would require logical changes to KIC loadbalancercontroller, mesh-api, cert-reloader, and our sidecar agent.

I think that the current approach is simplest and easiest to maintain.

Reply:
Oh I see. Those components don't have in-memory certs. Maybe slightly improving the constructor description for both writers will help clear things up then, just so we know the details of the differences between them, and their uses.